pkg/query/v2/logical/plan_indexscan.go

```diff
@@ -0,0 +1,244 @@
+// under the License.
+
+package logical
+
+import (
+	"bytes"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/pkg/errors"
+
+	commonv2 "github.com/apache/skywalking-banyandb/api/proto/banyandb/common/v2"
+	databasev2 "github.com/apache/skywalking-banyandb/api/proto/banyandb/database/v2"
+	streamv2 "github.com/apache/skywalking-banyandb/api/proto/banyandb/stream/v2"
+	"github.com/apache/skywalking-banyandb/banyand/tsdb"
+	"github.com/apache/skywalking-banyandb/pkg/index"
+	"github.com/apache/skywalking-banyandb/pkg/query/v2/executor"
+)
+
+var _ UnresolvedPlan = (*unresolvedIndexScan)(nil)
+
+type unresolvedIndexScan struct {
+	unresolvedOrderBy *UnresolvedOrderBy
+	startTime         time.Time
+	endTime           time.Time
+	metadata          *commonv2.Metadata
+	conditions        []Expr
+	projectionFields  [][]*Tag
+	entity            tsdb.Entity
+}
+
+func (uis *unresolvedIndexScan) Type() PlanType {
+	return PlanIndexScan
+}
+
+func (uis *unresolvedIndexScan) Analyze(s Schema) (Plan, error) {
+	conditionMap := make(map[*databasev2.IndexRule][]Expr)
+	for _, cond := range uis.conditions {
+		if resolvable, ok := cond.(ResolvableExpr); ok {
+			err := resolvable.Resolve(s)
+			if err != nil {
+				return nil, err
+			}
+
+			if bCond, ok := cond.(*binaryExpr); ok {
+				tag := bCond.l.(*FieldRef).tag
+				if defined, indexObj := s.IndexDefined(tag); defined {
+					if v, exist := conditionMap[indexObj]; exist {
+						v = append(v, cond)
+						conditionMap[indexObj] = v
+					} else {
+						conditionMap[indexObj] = []Expr{cond}
+					}
+				} else {
+					return nil, errors.Wrap(ErrIndexNotDefined, tag.GetCompoundName())
+				}
+			}
+		}
+	}
+
+	var projFieldsRefs [][]*FieldRef
+	if uis.projectionFields != nil && len(uis.projectionFields) > 0 {
+		var err error
+		projFieldsRefs, err = s.CreateRef(uis.projectionFields...)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// resolve sub-plan with the projected view of schema
+	orderBySubPlan, err := uis.unresolvedOrderBy.analyze(s.Proj(projFieldsRefs...))
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &indexScan{
+		orderBy:             orderBySubPlan,
+		timeRange:           tsdb.NewTimeRange(uis.startTime, uis.endTime),
+		schema:              s,
+		projectionFieldRefs: projFieldsRefs,
+		metadata:            uis.metadata,
+		conditionMap:        conditionMap,
+		entity:              uis.entity,
+	}, nil
+}
+
+var _ Plan = (*indexScan)(nil)
+
+type indexScan struct {
+	*orderBy
+	timeRange           tsdb.TimeRange
+	schema              Schema
+	metadata            *commonv2.Metadata
+	conditionMap        map[*databasev2.IndexRule][]Expr
+	projectionFieldRefs [][]*FieldRef
+	entity              tsdb.Entity
+}
+
+func (i *indexScan) Execute(ec executor.ExecutionContext) ([]*streamv2.Element, error) {
+	shards, err := ec.Shards(i.entity)
+	if err != nil {
+		return nil, err
+	}
+	var elements [][]*streamv2.Element
```

Comment:
No, we should not load all elements here.

The algorithm should be :

 1. Move all iterator to the first item by invoking their `Next`.
 2. Load all `first` items into a slice. 
 3. Pop up the minimal item through the order value, parse it to element, as the first value in the final result.
 4. Move the iterator whose value is poped in step 3, push the next value of this iterator into the slice.
 5. Repeat 3~4 until the final result reaches the limitation in the criteria.
 
This algorithm intends to use less space by loading elements that will be selected.